---
title: "grosbeak data investigation"
author: "Devon DeRaad"
date: "9/2/2021"
output: html_document
---

```{r}
#install and load necessary packages
#devtools::install_github("DevonDeRaad/SNPfiltR")
library(SNPfiltR)
library(vcfR)
library(StAMPP)
library(adegenet)
```

```{r}
#read in vcf as vcfR
vcfR <- read.vcfR("~/Desktop/grosbeaks/m3.vcf")

#generate popmap file. Two column popmap with the same format as stacks, and the columns must be named 'id' and 'pop'
popmap<-data.frame(id=colnames(vcfR@gt)[2:length(colnames(vcfR@gt))],pop=substr(colnames(vcfR@gt)[2:length(colnames(vcfR@gt))], 3,8))

#hard filter to minimum depth of 5, and minimum genotype quality of 30
vcfR<-hard_filter(vcfR=vcfR, depth = 5, gq = 30)
```

```{r}
#execute allele balance filter
vcfR<-filter_allele_balance(vcfR)
```

```{r}
#visualize and pick appropriate max depth cutoff
max_depth(vcfR)

#filter vcf by the max depth cutoff you chose
vcfR<-max_depth(vcfR, maxdepth = 100)

#check vcfR to see how many SNPs we have left
vcfR
```

```{r}
#run function to visualize missing data per sample
missing_by_sample(vcfR=vcfR, popmap = popmap)

#set missing data cutoff for individual samples
vcfR.95<-missing_by_sample(vcfR=vcfR, cutoff = .95)
```

```{r}
#subset popmap to only include retained individuals
popmap.95<-popmap[popmap$id %in% colnames(vcfR.95@gt),]

#remove invariant sites generated throughout the filtering process
vcfR.95<-min_mac(vcfR = vcfR.95, min.mac = 1)
```

```{r}
#visualize missing data by SNP and the effect of various cutoffs on the missingness of each sample
missing_by_snp(vcfR.95)
```

```{r}
#rerun missing by sample to figure out which samples are problematic even at high completeness cutoffs
miss<-missing_by_sample(vcfR=vcfR.95, popmap = popmap.95)
```

```{r}
#subset dataframe to show only missing data at .9 completeness cutoff
miss.9<-miss[miss$filt == .9,]
#histogram of samples
hist(miss.9$snps.retained, breaks=50)
#order the samples to figure out which to drop
miss.9[order(miss.9$snps.retained),]
#manually drop 3 worst samples which are still unusable at high missing data cutoffs
vcfR.95 <- vcfR.95[,colnames(vcfR.95@gt) != "P_melanocephalus_39985" & 
                     colnames(vcfR.95@gt) != "P_melanocephalus_44702" &
                     colnames(vcfR.95@gt) != "P_melanocephalus_44836"]

dim(vcfR.95)
```

```{r}
#remove invariant sites potentially generated by dropping samples
vcfR.95<-min_mac(vcfR = vcfR.95, min.mac = 1)

#subset popmap to only include retained individuals
popmap.95<-popmap[popmap$id %in% colnames(vcfR.95@gt),]
```

```{r}
#revisualize missing data by SNP and the effect of various cutoffs on the missingness of each sample
missing_by_snp(vcfR.95)
```

```{r}
#visualize the effect of a few reasonable potential missing data cutoffs on clustering patterns 
g<-assess_missing_data(vcfR=vcfR.95, popmap = popmap.95, thresholds= c(.85,.9,.95), iterations = 3000)
```

```{r}

vcfR.95.90<-missing_by_snp(vcfR.95, cutoff = .90)
vcfR.95.90

#distance thin
f<-distance_thin(vcfR = vcfR.95.90, min.distance = 500)
```

```{r}
#plot depth per snp and per sample
dp <- extract.gt(vcfR.95.90, element = "DP", as.numeric=TRUE)
heatmap.bp(dp, rlabels = FALSE)

#plot genotype quality per snp and per sample
gq <- extract.gt(vcfR.95.90, element = "GQ", as.numeric=TRUE)
heatmap.bp(gq, rlabels = FALSE)
```

```{r}
#convert vcfR to genlight
gen<-vcfR2genlight(vcfR.95.90)
### Calculate Nei's distances between individuals/pops
pop(gen)<-popmap.95$pop
sample.div <- stamppNeisD(gen, pop = FALSE)
#NJ tree
plot(nj(sample.div), type = "unrooted", cex = .65)
#export for splitstree
#stamppPhylip(distance.mat=sample.div, file="~/Downloads/grosbeak.splits.txt")
#very shallow divergence among samples and pops
stamppFst(gen, nboots = 1, percent = 95, nclusters = 1)

#vcfR::write.vcf(vcfR.95.90, "~/Desktop/grosbeaks/filtered.vcf.gz")
#vcfR::write.vcf(f, "~/Desktop/grosbeaks/filtered.unlinked.vcf.gz")
```



