---
title: "basic_clustering"
author: "Devon DeRaad"
date: "9/3/2021"
output: html_document
---

```{r}
#install and load necessary packages
#devtools::install_github("DevonDeRaad/SNPfiltR")
library(SNPfiltR)
library(vcfR)
library(StAMPP)
library(adegenet)
library(ggplot2)
library(introgress)
```

```{r}
#read vcfR
vcfR<-read.vcfR("~/Desktop/grosbeaks/grosbeaks.rad/filtered.vcf.gz")

#read locs
locs<-read.csv("~/Desktop/grosbeaks/grosbeaks.rad/grosbeak.sampling.scheme.csv")
#locs$species<-sub("._","",locs$sample)
#locs$species<-sub("_.*","",locs$species)
#write.csv(locs, "~/Desktop/grosbeaks/grosbeaks.rad/grosbeak.sampling.scheme.csv")
#retain only retained samples
locs<-locs[locs$sample %in% colnames(vcfR@gt)[-1],]
#match order between samples in vcf and locs
locs<-locs[match(colnames(vcfR@gt)[-1],locs$sample),]
```

```{r}
#convert vcfr to genlight
gen<- vcfR2genlight(vcfR)
#make pca
pca<-glPca(gen, nf=6)

#pull pca scores out of df
pca.scores<-as.data.frame(pca$scores)
#match order
pca.scores$sex<-locs$sex
pca.scores$state<-locs$state
pca.scores$species<-locs$species

#ggplot color by species
ggplot(pca.scores, aes(x=PC1, y=PC2, color=species)) +
  geom_point(cex = 4, alpha=.75)+
  theme_classic()+
  theme(legend.position = "bottom", legend.key.size = unit(0, 'cm'),
        legend.title = element_text(size=8), legend.text = element_text(size=6.5))

#ggplot color by sex
ggplot(pca.scores, aes(x=PC1, y=PC2, color=sex)) +
  geom_point(cex = 4, alpha=.75)+
  theme_classic()+
  theme(legend.position = "bottom", legend.key.size = unit(0, 'cm'),
        legend.title = element_text(size=8), legend.text = element_text(size=6.5))

#ggplot color by state
ggplot(pca.scores, aes(x=PC1, y=PC2, color=state)) +
  geom_point(cex = 4, alpha=.75)+
  theme_classic()+
  theme(legend.position = "bottom", legend.key.size = unit(0, 'cm'),
        legend.title = element_text(size=8), legend.text = element_text(size=6.5))

```

```{r}
#calculate missingness per sample
mat<-extract.gt(vcfR)
miss<-c()
for (i in 1:ncol(mat)){
  miss[i]<-sum(is.na(mat[,i]))
}
locs$miss<-miss
#locs 

#remove loci that aren't biallelic
mat<-mat[nchar(vcfR@fix[,"ALT"]) == 1,]
dim(mat)
#identify SNPs that are fixed away from LA
conv.mat<-mat
conv.mat[1:5,1:5]
conv.mat[conv.mat == "0/0"]<-0
conv.mat[conv.mat == "0/1"]<-1
conv.mat[conv.mat == "1/1"]<-2
conv.mat<-as.data.frame(conv.mat)
#convert to numeric
for (i in 1:ncol(conv.mat)){
  conv.mat[,i]<-as.numeric(as.character(conv.mat[,i]))
}

#show colnames to verify you're subsetting correctly
colnames(conv.mat) 
#calc AF
mel.af<-(rowSums(conv.mat[,c(2,7,10,15,24,75,78,88,99,128,142)], na.rm=T)/(rowSums(is.na(conv.mat[,c(2,7,10,15,24,75,78,88,99,128,142)]) == FALSE)))/2
lud.af<-(rowSums(conv.mat[,c(139,122,92,87,50)], na.rm=T)/(rowSums(is.na(conv.mat[,c(139,122,92,87,50)]) == FALSE)))/2

#find fixed SNPs
diff<-abs(mel.af - lud.af)
table(is.na(diff))
#how many SNPs are fixed
table(is.na(diff) == FALSE & diff == 1)

#list some fixed SNPs
head(vcfR@fix[,1][is.na(diff) == FALSE & diff == 1])

#subsample original matrix to only fixed diff SNPs
gen.mat<-mat[is.na(diff) == FALSE & diff == 1,]
dim(gen.mat)
#subsample matrix converted for AF calcs to only fixed SNPS
conv.mat<-conv.mat[is.na(diff) == FALSE & diff == 1,]
dim(conv.mat)
#write a logical test to convert alleles so that a single number represents one parental ancestry
for (i in 1:nrow(gen.mat)){
  #if 1 is the melano allele
  if((sum(conv.mat[i,c(2,7,10,15,24,75,78,88,99,128,142)], na.rm=T)/(sum(is.na(conv.mat[i,c(2,7,10,15,24,75,78,88,99,128,142)]) == FALSE)))/2 == 0){
    #swap all '0/0' cells with '2/2'
    gen.mat[i,][gen.mat[i,] == "0/0"]<-"2/2"
    #swap all '1/1' cells with '0/0'
    gen.mat[i,][gen.mat[i,] == "1/1"]<-"0/0"
    #finally convert all '2/2' cells (originally 0/0) into '1/1'
    gen.mat[i,][gen.mat[i,] == "2/2"]<-"1/1"
    #no need to touch hets
  }
}

#convert R class NAs to the string "NA/NA"
gen.mat[is.na(gen.mat) == TRUE]<-"NA/NA"

#if it worked correctly this should be only missing or '1/1'
table(gen.mat[,7])
table(gen.mat[,2])
table(gen.mat[,10])

#make locus info df
locus.info<-data.frame(locus=rownames(gen.mat),
                       type=rep("C", times=nrow(gen.mat)),
                       lg=gsub("_.*","",rownames(gen.mat)),
                       marker.pos=gsub(".*_","",rownames(gen.mat)))
#make linkage group numeric
locus.info$lg<-as.numeric(as.character(locus.info$lg))
locus.info$marker.pos<-as.numeric(as.character(locus.info$marker.pos))

#we now have a gt matrix in proper format for introgress
#convert genotype data into a matrix of allele counts
count.matrix<-prepare.data(admix.gen=gen.mat, loci.data=locus.info,
                           parental1="1",parental2="0", pop.id=F,
                           ind.id=F, fixed=T)

#estimate hybrid index values
hi.index.sim<-est.h(introgress.data=count.matrix,loci.data=locus.info,
                    fixed=T, p1.allele="1", p2.allele="0")
hi.index.sim$sample<-locs$sample
hi.index.sim$state<-locs$state
hi.index.sim$sex<-locs$sex

#locus.info$locus<-rep("", times=nrow(locus.info))
#LociDataSim1$lg<-c(1:110)
mk.image(introgress.data=count.matrix, loci.data=locus.info,
         marker.order=NULL,hi.index=hi.index.sim, ylab.image="Individuals",
         xlab.h="Red-crowned ancestry", pdf=F,
         col.image=c("red","green","blue"))

```

```{r}
#calculate mean heterozygosity
het<-calc.intersp.het(introgress.data=count.matrix)

#make triangle plot colored by species
plot(x=hi.index.sim$h, y=het, bg=rgb(0,0,0,alpha=0.3), pch=21, cex=2, col=as.factor(locs$species),
     xlab="Hybrid Index", ylab="Interspecific heterozygosity",
     ylim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)
#red = ludovicianus, green = melanocephalus, black = hybrid

#make triangle plot colored by species
plot(x=hi.index.sim$h, y=het, bg=rgb(0,0,0,alpha=0.3), pch=21, cex=2, col=as.factor(locs$state),
     xlab="Hybrid Index", ylab="Interspecific heterozygosity",
     ylim=c(0,1))
segments(x0 =0, y0 =0, x1 =.5, y1 =1)
segments(x0 =1, y0 =0, x1 =.5, y1 =1)
#blue indicates from the SD transect, red indicates Kansas/missouri

```

